docker.go

```diff
@@ -628,26 +628,44 @@ func (c *DockerContainer) StartLogProducer(ctx context.Context) error {
 			Since:      since,
 		}
 
-		ctx, cancel := context.WithTimeout(ctx, time.Second*5)
-		defer cancel()
-
 		r, err := c.provider.client.ContainerLogs(ctx, c.GetContainerID(), options)
 		if err != nil {
-			// if we can't get the logs, panic, we can't return an error to anything
-			// from within this goroutine
-			panic(err)
+			// if we can't get the logs, retry in one second.
+			if ctx.Err() != nil {
+				// context done.
+				return
+			}
+
+			// we need a separate select here in order to be able to print the error
+			// and this cannot be done in the same select that checks for the timeout.
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				c.logger.Printf("cannot get logs for container %q: %v", c.ID, err)
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(1 * time.Second):
+			}
+
+			goto BEGIN
 		}
 		defer c.provider.Close()
 
 		for {
-			select {
-			case <-stop:
+			if ctx.Err() != nil {
 				err := r.Close()
 				if err != nil {
 					// we can't close the read closer, this should never happen
 					panic(err)
 				}
-				return
+			}
```

Comment:
You might move this whole `if` into the `select` case in lines 667-668 where you check if the context is done.